Move async article from WebFu. (#4450)
GoogleChrome/web.dev: src/site/content/en/blog/async-functions/index.md

description: >
  Async functions allow you to write promise-based code as if it were synchronous.
date: 2016-10-20
updated: 2021-01-18
tags:
  - javascript
authors:
  - jakearchibald
feedback:
  - api
---

Async functions were enabled by default in Chrome 55 and they're quite frankly
marvelous. They allow you to write promise-based code as if it were synchronous,
but without blocking the main thread. They make your asynchronous code less
"clever" and more readable. If you're unfamilliar with promises, you might want to checkout [JavaScript Promises: an introduction](./promises).

Async functions work like this:

```js
async function myFirstAsyncFunction() {
  try {
    const fulfilledValue = await promise;
  }
  catch (rejectedValue) {
    // …
  }
}
```

If you use the `async` keyword before a function definition, you can then use
`await` within the function. When you `await` a promise, the function is paused
in a non-blocking way until the promise settles. If the promise fulfills, you
get the value back. If the promise rejects, the rejected value is thrown.

{% Aside %}
If you're unfamiliar with promises, check out [our

> promises guide](/promises).

Reviewer: Does this link work? It's `./promises` above.

{% endAside %}

## Example: logging a fetch

Say you want to fetch a URL and log the response as text. Here's how it looks
using promises:

```js
function logFetch(url) {
  return fetch(url)
    .then(response => response.text())
    .then(text => {
      console.log(text);
    }).catch(err => {
      console.error('fetch failed', err);
    });
}
```

And here's the same thing using async functions:

```js
async function logFetch(url) {
  try {
    const response = await fetch(url);
    console.log(await response.text());
  }
  catch (err) {
    console.log('fetch failed', err);
  }
}
```

It's the same number of lines, but all the callbacks are gone. This makes it way
easier to read, especially for those less familiar with promises.

{% Aside %}
Anything you `await` is passed through `Promise.resolve()`, so you can

> safely `await` non-native promises.

Reviewer 2: @jpmedley please rephrase to avoid "native"

Author: @jakearchibald I don't know what non-native promises are.

Reviewer: Added suggestion below

    return reader.read().then(function processResult(result) {
      if (result.done) return total;

      const value = result.value;
      total += value.length;
      console.log('Received chunk', value);

      return reader.read().then(processResult);
    })
  });
}
```

Check me out, Jake "wielder of promises" Archibald. See how I'm calling
`processResult()` inside itself to set up an asynchronous loop? Writing that made
me feel *very smart*. But like most "smart" code, you have to stare at it for
ages to figure out what it's doing, like one of those magic-eye pictures from
the 90's.

Let's try that again with async functions:

```js
async function getResponseSize(url) {
  const response = await fetch(url);
  const reader = response.body.getReader();
  let result = await reader.read();
  let total = 0;

  while (!result.done) {
    const value = result.value;
    total += value.length;
    console.log('Received chunk', value);
    // get the next result
    result = await reader.read();
  }

  return total;
}
```

All the "smart" is gone. The asynchronous loop that made me feel so smug is
replaced with a trusty, boring, while-loop. Much better. In future, you'll get

> [async iterators](https://github.com/tc39/proposal-async-iteration){: .external},

Reviewer 2: ```suggestion
[async iterators](https://github.com/tc39/proposal-async-iteration),
```
Please remove all these `{: .external}` things. They don't do anything on web.dev and they are breaking the build.

Author: Removed.

which would
[replace the `while` loop with a for-of loop](https://gist.github.com/jakearchibald/0b37865637daf884943cf88c2cba1376){: .external}, making it even neater.

{% Aside %}
I'm sort-of in love with streams. If you're unfamiliar with streaming,
[check out my guide](https://jakearchibald.com/2016/streams-ftw/#streams-the-fetch-api){: .external}.
{% endAside %}

## Other async function syntax

> I'ver shown you `async function() {}` already, but the `async` keyword can be

Reviewer: ```suggestion
I've shown you `async function() {}` already, but the `async` keyword can be
```

  const wait1 = wait(500); // Start a 500ms timer asynchronously…
  const wait2 = wait(500); // …meaning this timer happens in parallel.
  await wait1; // Wait 500ms for the first timer…
  await wait2; // …by which time this timer has already finished.
  return "done!";
}
```

The above takes 500ms to complete, because both waits happen at the same time.
Let's look at a practical example.

### Example: outputting fetches in order

Say you wanted to fetch a series of URLs and log them as soon as possible, in the
correct order.

*Deep breath* - here's how that looks with promises:

```js
function logInOrder(urls) {
  // fetch all the URLs
  const textPromises = urls.map(url => {
    return fetch(url).then(response => response.text());
  });

  // log them in order
  textPromises.reduce((chain, textPromise) => {
    return chain.then(() => textPromise)
      .then(text => console.log(text));
  }, Promise.resolve());
}
```

Yeah, that's right, I'm using `reduce` to chain a sequence of promises. I'm *so
smart*. But this is a bit of *so smart* coding you're better off without.

However, when converting the above to an async function, it's tempting to go
*too sequential*:

> <span class="compare-worse">Not recommended</span> - too sequential

Reviewer 2: Please replace https://web.dev/handbook/web-dev-components/#compare

Author: Done.

```js
async function logInOrder(urls) {
  for (const url of urls) {
    const response = await fetch(url);
    console.log(await response.text());
  }
}
```

Looks much neater, but my second fetch doesn't begin until my first fetch has
been fully read, and so on. This is much slower than the promises example that
performs the fetches in parallel. Thankfully there's an ideal middle-ground:

> <span class="compare-better">Recommended</span> - nice and parallel

Reviewer 2: Please replace https://web.dev/handbook/web-dev-components/#compare

Author: Done.

```js
async function logInOrder(urls) {
  // fetch all the URLs in parallel
  const textPromises = urls.map(async url => {
    const response = await fetch(url);
    return response.text();
  });

  // log them in sequence
  for (const textPromise of textPromises) {
    console.log(await textPromise);
  }
}
```

In this example, the URLs are fetched and read in parallel, but the "smart"
`reduce` bit is replaced with a standard, boring, readable for-loop.

## Browser support and workarounds

At the time of writing, async functions are enabled by default in Chrome, Edge,

> Firefox, and Safari.

Reviewer 2: Please move this so that it's the first section of the page

Author: Changed the first line to:

>Async functions are enabled by default in Chrome, Edge, Firefox, and Safari, and they're quite frankly marvelous. 

Changed this paragraph to:

>Async functions are enabled by default in most major browsers. If your site still needs to support older browsers without async, here are a few workarounds.

### Workaround: generators

If you're targeting browsers that support generators (which includes

> [the latest version of every major browser](http://kangax.github.io/compat-table/es6/#test-generators){:.external}

Reviewer 3: Netlify is failing to build because somewhere in this file, one of these `{: }` directives isn't working. Taking a quick glance I was wondering if it's this one because it doesn't have a space between the colon and the dot?

Author: I tried adding a space. See if that works.

Author: Everything seems to pass now. PTAL.

    // …
  }
}
```

If you use the `async` keyword before a function definition, you can then use
`await` within the function. When you `await` a promise, the function is paused
in a non-blocking way until the promise settles. If the promise fulfills, you
get the value back. If the promise rejects, the rejected value is thrown.

{% Aside %}
If you're unfamiliar with promises, check out [our
promises guide](/promises).
{% endAside %}

## Example: logging a fetch

Say you want to fetch a URL and log the response as text. Here's how it looks
using promises:

```js
function logFetch(url) {
  return fetch(url)
    .then(response => response.text())
    .then(text => {
      console.log(text);
    }).catch(err => {
      console.error('fetch failed', err);
    });
}
```

And here's the same thing using async functions:

```js
async function logFetch(url) {
  try {
    const response = await fetch(url);
    console.log(await response.text());
  }
  catch (err) {
    console.log('fetch failed', err);
  }
}
```

It's the same number of lines, but all the callbacks are gone. This makes it way
easier to read, especially for those less familiar with promises.

{% Aside %}

> Anything you `await` is passed through `Promise.resolve()`, so you can
> safely `await` non-native promises.

Reviewer: ```suggestion
Anything you `await` is passed through `Promise.resolve()`, so you can
safely `await` non-platform promises, such as those created by promise polyfills.
```
